File: typescript/packages/mechanisms/evm/src/erc4337/bundlers/pimlico.ts
/**
 * Pimlico Bundler Client
 *
 * Extended bundler client with Pimlico-specific methods:
 * - pimlico_getUserOperationGasPrice: Get current gas prices
 * - pimlico_sendCompressedUserOperation: Submit with calldata compression
 *
 * @see https://docs.pimlico.io/bundler/reference
 */

import type { Address, Hex } from "viem";
import {
  BundlerClient,
  BundlerError,
} from "../bundler.js";
import type { BundlerConfig } from "../types.js";
import type {
  UserOperation,
  UserOperationResult,
  GasEstimate,
} from "../types.js";
import { ENTRYPOINT_V07_ADDRESS, packAccountGasLimits, packGasFees } from "../constants.js";

/**
 * Gas price response from Pimlico
 */
export interface PimlicoGasPrice {
  slow: {
    maxFeePerGas: bigint;
    maxPriorityFeePerGas: bigint;
  };
  standard: {
    maxFeePerGas: bigint;
    maxPriorityFeePerGas: bigint;
  };
  fast: {
    maxFeePerGas: bigint;
    maxPriorityFeePerGas: bigint;
  };
}

/**
 * Pimlico bundler configuration
 */
export interface PimlicoConfig {
  /** Bundler RPC URL (optional, auto-generated if not provided) */
  bundlerUrl?: string;
  /** Chain ID */
  chainId: number;
  /** Pimlico API key */
  apiKey: string;
}

/**
 * Pimlico bundler client with extended methods
 */
export class PimlicoBundlerClient extends BundlerClient {
  private readonly apiKey: string;
  private readonly pimlicoUrl: string;

  constructor(config: PimlicoConfig) {
    // Construct Pimlico bundler URL
    const bundlerUrl = config.bundlerUrl && config.bundlerUrl.includes("pimlico")
      ? config.bundlerUrl
      : `https://api.pimlico.io/v2/${config.chainId}/rpc?apikey=${config.apiKey}`;

    super({
      ...config,
      bundlerUrl,
    });

    this.apiKey = config.apiKey;
    this.pimlicoUrl = bundlerUrl;
  }

  /**
   * Get current gas prices from Pimlico
   * Returns slow, standard, and fast gas price estimates
   */
  async getUserOperationGasPrice(): Promise<PimlicoGasPrice> {
    const result = await this.pimlicoRpcCall<{
      slow: { maxFeePerGas: Hex; maxPriorityFeePerGas: Hex };
      standard: { maxFeePerGas: Hex; maxPriorityFeePerGas: Hex };
      fast: { maxFeePerGas: Hex; maxPriorityFeePerGas: Hex };
    }>("pimlico_getUserOperationGasPrice", []);

    return {
      slow: {
        maxFeePerGas: BigInt(result.slow.maxFeePerGas),
        maxPriorityFeePerGas: BigInt(result.slow.maxPriorityFeePerGas),
      },
      standard: {
        maxFeePerGas: BigInt(result.standard.maxFeePerGas),
        maxPriorityFeePerGas: BigInt(result.standard.maxPriorityFeePerGas),
      },
      fast: {
        maxFeePerGas: BigInt(result.fast.maxFeePerGas),
        maxPriorityFeePerGas: BigInt(result.fast.maxPriorityFeePerGas),
      },
    };
  }

  /**
   * Send a compressed UserOperation for lower gas costs
   * Pimlico compresses the calldata before submitting to reduce L1 costs
   */
  async sendCompressedUserOperation(
    userOp: UserOperation,
    inflatorAddress: Address,
  ): Promise<UserOperationResult> {
    const packed = this.packUserOpForRpc(userOp);

    const userOpHash = await this.pimlicoRpcCall<Hex>(
      "pimlico_sendCompressedUserOperation",
      [packed, inflatorAddress, ENTRYPOINT_V07_ADDRESS],
    );

    return {
      userOpHash,
      wait: () => this.waitForReceipt(userOpHash),
    };
  }

  /**
   * Get user operation status from Pimlico
   */
  async getUserOperationStatus(userOpHash: Hex): Promise<{
    status: "not_found" | "not_submitted" | "submitted" | "rejected" | "included" | "failed" | "reverted";
    transactionHash?: Hex;
  }> {
    const result = await this.pimlicoRpcCall<{
      status: string;
      transactionHash?: Hex;
    }>("pimlico_getUserOperationStatus", [userOpHash]);

    return {
      status: result.status as "not_found" | "not_submitted" | "submitted" | "rejected" | "included" | "failed" | "reverted",
      transactionHash: result.transactionHash,
    };
  }

  /**
   * Estimate gas with Pimlico-specific optimizations
   */
  async estimateUserOperationGasWithPimlico(
    userOp: Partial<UserOperation> & {
      sender: Address;
      callData: Hex;
    },
    stateOverride?: Record<Address, { balance?: Hex; code?: Hex; nonce?: Hex }>,
  ): Promise<GasEstimate> {
    const estimationOp = {
      sender: userOp.sender,
      nonce: this.bigintToHex(userOp.nonce ?? 0n),
      initCode: userOp.initCode ?? "0x",
      callData: userOp.callData,
      accountGasLimits: packAccountGasLimits(
        userOp.verificationGasLimit ?? 1000000n,
        userOp.callGasLimit ?? 1000000n,
      ),
      preVerificationGas: this.bigintToHex(userOp.preVerificationGas ?? 100000n),
      gasFees: packGasFees(
        userOp.maxPriorityFeePerGas ?? 1000000000n,
        userOp.maxFeePerGas ?? 10000000000n,
      ),
      paymasterAndData: userOp.paymasterAndData ?? "0x",
      signature:
        userOp.signature ??
        "0xfffffffffffffffffffffffffffffff0000000000000000000000000000000007aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa1c",
    };

    const params: unknown[] = [estimationOp, ENTRYPOINT_V07_ADDRESS];
    if (stateOverride) {
      params.push(stateOverride);
    }

    const result = await this.pimlicoRpcCall<{
      verificationGasLimit: Hex;
      callGasLimit: Hex;
      preVerificationGas: Hex;
      paymasterVerificationGasLimit?: Hex;
      paymasterPostOpGasLimit?: Hex;
    }>("eth_estimateUserOperationGas", params);

    return {
      verificationGasLimit: BigInt(result.verificationGasLimit),
      callGasLimit: BigInt(result.callGasLimit),
      preVerificationGas: BigInt(result.preVerificationGas),
      paymasterVerificationGasLimit: result.paymasterVerificationGasLimit
        ? BigInt(result.paymasterVerificationGasLimit)
        : undefined,
      paymasterPostOpGasLimit: result.paymasterPostOpGasLimit
        ? BigInt(result.paymasterPostOpGasLimit)
        : undefined,
    };
  }

  /**
   * Pack UserOperation for RPC
   */
  private packUserOpForRpc(userOp: UserOperation): Record<string, unknown> {
    return {
      sender: userOp.sender,
      nonce: this.bigintToHex(userOp.nonce),
      initCode: userOp.initCode,
      callData: userOp.callData,
      accountGasLimits: packAccountGasLimits(
        userOp.verificationGasLimit,
        userOp.callGasLimit,
      ),
      preVerificationGas: this.bigintToHex(userOp.preVerificationGas),
      gasFees: packGasFees(userOp.maxPriorityFeePerGas, userOp.maxFeePerGas),
      paymasterAndData: userOp.paymasterAndData,
      signature: userOp.signature,
    };
  }

  /**
   * Convert bigint to hex
   */
  private bigintToHex(value: bigint): Hex {
    return `0x${value.toString(16)}` as Hex;
  }

  /**
   * Make Pimlico-specific RPC call
   */
  private async pimlicoRpcCall<T>(method: string, params: unknown[]): Promise<T> {
    const response = await fetch(this.pimlicoUrl, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        jsonrpc: "2.0",
        id: Date.now(),
        method,
        params,
      }),
    });

    if (!response.ok) {
      throw new BundlerError(
        `HTTP error: ${response.status} ${response.statusText}`,
      );
    }

    const json = await response.json() as {
      result?: T;
      error?: { code: number; message: string; data?: unknown };
    };

    if (json.error) {
      throw new BundlerError(json.error.message, json.error.code, json.error.data);
    }

    return json.result as T;
  }
}

/**
 * Create a Pimlico bundler client
 */
export function createPimlicoBundlerClient(config: PimlicoConfig): PimlicoBundlerClient {
  return new PimlicoBundlerClient(config);
}

File: go/mechanisms/evm/erc4337/pimlico.go
package erc4337

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"math/big"
	"net/http"
	"time"

	"github.com/ethereum/go-ethereum/common"
)

// PimlicoConfig contains configuration for Pimlico bundler client.
type PimlicoConfig struct {
	// APIKey is the Pimlico API key
	APIKey string `json:"apiKey"`
	// ChainID is the chain ID
	ChainID int64 `json:"chainId"`
	// BundlerURL is an optional custom bundler URL
	BundlerURL string `json:"bundlerUrl,omitempty"`
	// EntryPoint is the EntryPoint contract address (optional, defaults to v0.7)
	EntryPoint common.Address `json:"entryPoint,omitempty"`
}

// PimlicoGasPrice contains gas price estimates from Pimlico.
type PimlicoGasPrice struct {
	Slow struct {
		MaxFeePerGas         *big.Int `json:"maxFeePerGas"`
		MaxPriorityFeePerGas *big.Int `json:"maxPriorityFeePerGas"`
	} `json:"slow"`
	Standard struct {
		MaxFeePerGas         *big.Int `json:"maxFeePerGas"`
		MaxPriorityFeePerGas *big.Int `json:"maxPriorityFeePerGas"`
	} `json:"standard"`
	Fast struct {
		MaxFeePerGas         *big.Int `json:"maxFeePerGas"`
		MaxPriorityFeePerGas *big.Int `json:"maxPriorityFeePerGas"`
	} `json:"fast"`
}

// PimlicoBundlerClient is a Pimlico-specific bundler client.
type PimlicoBundlerClient struct {
	*GenericBundlerClient
	apiKey     string
	pimlicoURL string
}

// NewPimlicoBundlerClient creates a new Pimlico bundler client.
func NewPimlicoBundlerClient(config PimlicoConfig) *PimlicoBundlerClient {
	bundlerURL := config.BundlerURL
	if bundlerURL == "" {
		network := getPimlicoNetwork(config.ChainID)
		bundlerURL = fmt.Sprintf("https://api.pimlico.io/v2/%s/rpc?apikey=%s", network, config.APIKey)
	}

	entryPoint := config.EntryPoint
	if entryPoint == (common.Address{}) {
		entryPoint = common.HexToAddress(EntryPointV07Address)
	}

	genericClient := NewBundlerClient(BundlerConfig{
		BundlerURL: bundlerURL,
		EntryPoint: entryPoint,
		ChainID:    config.ChainID,
	})

	return &PimlicoBundlerClient{
		GenericBundlerClient: genericClient,
		apiKey:               config.APIKey,
		pimlicoURL:           bundlerURL,
	}
}

// GetUserOperationGasPrice retrieves gas prices from Pimlico.
func (c *PimlicoBundlerClient) GetUserOperationGasPrice() (*PimlicoGasPrice, error) {
	var result struct {
		Slow struct {
			MaxFeePerGas         string `json:"maxFeePerGas"`
			MaxPriorityFeePerGas string `json:"maxPriorityFeePerGas"`
		} `json:"slow"`
		Standard struct {
			MaxFeePerGas         string `json:"maxFeePerGas"`
			MaxPriorityFeePerGas string `json:"maxPriorityFeePerGas"`
		} `json:"standard"`
		Fast struct {
			MaxFeePerGas         string `json:"maxFeePerGas"`
			MaxPriorityFeePerGas string `json:"maxPriorityFeePerGas"`
		} `json:"fast"`
	}

	err := c.pimlicoRPCCall("pimlico_getUserOperationGasPrice", []interface{}{}, &result)
	if err != nil {
		return nil, err
	}

	return &PimlicoGasPrice{
		Slow: struct {
			MaxFeePerGas         *big.Int `json:"maxFeePerGas"`
			MaxPriorityFeePerGas *big.Int `json:"maxPriorityFeePerGas"`
		}{
			MaxFeePerGas:         hexToBigInt(result.Slow.MaxFeePerGas),
			MaxPriorityFeePerGas: hexToBigInt(result.Slow.MaxPriorityFeePerGas),
		},
		Standard: struct {
			MaxFeePerGas         *big.Int `json:"maxFeePerGas"`
			MaxPriorityFeePerGas *big.Int `json:"maxPriorityFeePerGas"`
		}{
			MaxFeePerGas:         hexToBigInt(result.Standard.MaxFeePerGas),
			MaxPriorityFeePerGas: hexToBigInt(result.Standard.MaxPriorityFeePerGas),
		},
		Fast: struct {
			MaxFeePerGas         *big.Int `json:"maxFeePerGas"`
			MaxPriorityFeePerGas *big.Int `json:"maxPriorityFeePerGas"`
		}{
			MaxFeePerGas:         hexToBigInt(result.Fast.MaxFeePerGas),
			MaxPriorityFeePerGas: hexToBigInt(result.Fast.MaxPriorityFeePerGas),
		},
	}, nil
}

// SendCompressedUserOperation sends a compressed UserOperation for gas savings.
func (c *PimlicoBundlerClient) SendCompressedUserOperation(
	compressedCalldata []byte,
	inflatorAddress common.Address,
) (common.Hash, error) {
	var result string
	err := c.pimlicoRPCCall("pimlico_sendCompressedUserOperation", []interface{}{
		bytesToHex(compressedCalldata),
		inflatorAddress.Hex(),
		c.entryPoint.Hex(),
	}, &result)
	if err != nil {
		return common.Hash{}, err
	}

	return common.HexToHash(result), nil
}

// GetUserOperationStatus gets the status of a UserOperation.
func (c *PimlicoBundlerClient) GetUserOperationStatus(userOpHash common.Hash) (*UserOperationStatus, error) {
	var result struct {
		Status        string `json:"status"`
		TransactionHash string `json:"transactionHash,omitempty"`
	}

	err := c.pimlicoRPCCall("pimlico_getUserOperationStatus", []interface{}{userOpHash.Hex()}, &result)
	if err != nil {
		return nil, err
	}

	status := &UserOperationStatus{
		Status: result.Status,
	}
	if result.TransactionHash != "" {
		status.TransactionHash = common.HexToHash(result.TransactionHash)
	}

	return status, nil
}

// UserOperationStatus represents the status of a UserOperation.
type UserOperationStatus struct {
	// Status: "not_found", "submitted", "pending", "included", "failed"
	Status          string      `json:"status"`
	TransactionHash common.Hash `json:"transactionHash,omitempty"`
}

// pimlicoRPCCall makes a Pimlico-specific RPC call.
func (c *PimlicoBundlerClient) pimlicoRPCCall(method string, params []interface{}, result interface{}) error {
	c.requestID++

	request := map[string]interface{}{
		"jsonrpc": "2.0",
		"id":      c.requestID,
		"method":  method,
		"params":  params,
	}

	body, err := json.Marshal(request)
	if err != nil {
		return fmt.Errorf("failed to marshal request: %w", err)
	}

	httpClient := &http.Client{Timeout: 30 * time.Second}
	resp, err := httpClient.Post(c.pimlicoURL, "application/json", bytes.NewReader(body))
	if err != nil {
		return fmt.Errorf("HTTP request failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("HTTP error %d: %s", resp.StatusCode, string(body))
	}

	var response struct {
		Result json.RawMessage `json:"result"`
		Error  *struct {
			Code    int             `json:"code"`
			Message string          `json:"message"`
			Data    json.RawMessage `json:"data,omitempty"`
		} `json:"error"`
	}

	if err := json.NewDecoder(resp.Body).Decode(&response); err != nil {
		return fmt.Errorf("failed to decode response: %w", err)
	}

	if response.Error != nil {
		return fmt.Errorf("RPC error %d: %s", response.Error.Code, response.Error.Message)
	}

	if result != nil && len(response.Result) > 0 {
		if err := json.Unmarshal(response.Result, result); err != nil {
			return fmt.Errorf("failed to unmarshal result: %w", err)
		}
	}

	return nil
}

// getPimlicoNetwork returns the Pimlico network name for a chain ID.
func getPimlicoNetwork(chainID int64) string {
	networks := map[int64]string{
		1:        "ethereum",
		11155111: "sepolia",
		137:      "polygon",
		80001:    "mumbai",
		10:       "optimism",
		420:      "optimism-goerli",
		42161:    "arbitrum",
		421613:   "arbitrum-goerli",
		8453:     "base",
		84532:    "base-sepolia",
	}

	if network, ok := networks[chainID]; ok {
		return network
	}
	return fmt.Sprintf("%d", chainID)
}
